IIP-0013: Dynamic Module Metadata (#48)
iotaledger/IIPs: iips/IIP-0013/IIP-0013.md

---
iip: 13
title: Dynamic Module Metadata
description: Store package module metadata in dynamic fields so it can be extended without changing the package metadata object type.
author: Lorenzo Benetollo (@lollobene) <lorenzo.benetollo@iota.org>, Mirko Zichichi (@miker83z) <mirko.zichichi@iota.org>
discussions-to: https://github.com/iotaledger/IIPs/discussions/47
status: Draft
type: Standards Track
layer: Core
created: 2026-06-17
requires: IIP-0010, IIP-0011
---

## Abstract

> This proposal refines the `PackageMetadata` model defined in [IIP-0010](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0010/IIP-0010.md) by moving per-module metadata out of the immutable `PackageMetadataV1` object and into **dynamic fields**. Instead of embedding an inline `VecMap<String, ModuleMetadataV1>`, the `PackageMetadataV1` object now records a metadata-layout version and references a map of `ModuleMetadata` objects, each derived deterministically from the package id and module name. Each `ModuleMetadata` is itself an extensible key-value store backed by dynamic fields, where every key holds a distinct kind of module metadata.

Reviewer: ```suggestion
This proposal refines the `PackageMetadata` model defined in [IIP-0010](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0010/IIP-0010.md) by moving per-module metadata out of the immutable `PackageMetadataV1` object and into **dynamic fields**. Instead of embedding modules' metadata into the `PackageMetadataV1`'s field `modules_metadata: VecMap<String, ModuleMetadataV1>`, a new set of dynamic fields is added to the `PackageMetadataV1` object for each module, i.e., a `ModuleMetadata` object.  Each `ModuleMetadata` is itself an extensible key-value store backed by dynamic fields, where every key holds a distinct kind of module metadata entry.
```

This proposal addresses both drawbacks by representing module metadata as **dynamic fields keyed by metadata kind**. New kinds are added by writing a new key; readers fetch only the keys they need; and the `PackageMetadata` object type stays stable.

## Specification

This proposal modifies only the **publish/upgrade object-creation phase** and the **runtime accessors** of the Package Metadata model. The compilation phase, the trust model, and the package-metadata object id derivation are unchanged from [IIP-0010](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0010/IIP-0010.md): metadata is still extracted from verified bytecode by the protocol, the object is still frozen, and its id is still derived from the package storage id.

### Layout

> When the relevant feature flag is enabled (see [Activation](#activation)), the protocol builds `PackageMetadataV1` with the following dynamic-field layout instead of the inline `modules_metadata` map.

Reviewer: ```suggestion
The protocol builds `PackageMetadataV1` with the following dynamic-field layout instead of the inline `modules_metadata` map.
```
I would avoid protocol flag related information. This is more of an implementation detail rather than part of the specification.

> The `PackageMetadataV1` object carries two dynamic fields:

Reviewer: ```suggestion
When the relevant feature flag is enabled (see [Activation](#activation)), the protocol builds `PackageMetadataV1` with the following dynamic-field layout:
```

> 1. A **metadata-layout version** marker under the `PackageMetadataVersionFieldName` key, set to `2`. Its presence and value distinguish the dynamic-field layout from the legacy inline layout.

Reviewer: ```suggestion
1. A **metadata-layout version** marker under the `PackageMetadataVersionFieldName` key (starting from `2`, in order to distinguish from the implementation proposed for [IIP-0010](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0010/IIP-0010.md)). Its presence allows to understand which layout a developer must reference to.
```

2. A **modules map** under the `ModulesMetadataFieldName` key, of type `VecMap<ModuleName, ModuleMetadata>`, mapping each module name to its `ModuleMetadata` object.

> The inline `modules_metadata: VecMap<ascii::String, ModuleMetadataV1>` field of `PackageMetadataV1` remains in the struct definition for backwards compatibility but is left **empty** under this layout; it is deprecated and MUST NOT be read by new code.

Reviewer: ```suggestion
The inline `modules_metadata: VecMap<ascii::String, ModuleMetadataV1>` field of `PackageMetadataV1` is left empty under this layout; it is deprecated and MUST NOT be read by new code (it is kept in the struct definition for backwards compatibility).
```

- `ModuleMetadataV1FieldName` → `ModuleMetadataV1` — the authenticator metadata from IIP-0010, re-homed under this key. Present only for modules that have at least one authenticator.
- `ViewFunctionMetadataV1FieldName` → `vector<ascii::String>` — the list of view function names declared in the module (new; see [View function metadata](#view-function-metadata)). Present for every module recorded under this layout.

Additional kinds can be introduced later as new key types without changing `ModuleMetadata`, `PackageMetadataV1`, or any existing key's payload.

### Module metadata object id derivation

Each `ModuleMetadata` object id is derived deterministically from the owning package storage id and the module name, reusing the same derived-object mechanism IIP-0010 uses for the package metadata object id:

```rust
module_metadata_id
= derive_object_id(package_storage_id, <0x2::module_metadata::ModuleMetadataKey>, module_name)
```

Where `ModuleMetadataKey` wraps the module name (`ModuleMetadataKey(ascii::String)`). Deriving from the package id together with the module name guarantees that:

> - each module of a package gets a distinct `ModuleMetadata` object (and therefore a distinct dynamic-field namespace), and

Reviewer: ```suggestion
- each module of a package gets a distinct `ModuleMetadata` object, and
```

- the same module name in different packages does not collide.

> As with the package metadata id, this id can be computed on-chain and by off-chain tooling without a lookup.

Reviewer: ```suggestion
As with the package metadata id, this approach has the advantage that the id can be computed by off-chain tooling without a lookup.
```

### View function metadata

A function annotated with `#[view]` ([IIP-0011](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0011/IIP-0011.md)) is recorded in its module's `ModuleMetadata` under the `ViewFunctionMetadataV1FieldName` key as a `vector<ascii::String>` of function names.

The flow mirrors the authenticator flow of IIP-0010:

1. **Compilation**: the `#[view]` attribute is recorded in the function's `RuntimeModuleMetadata` and embedded in the module bytecode, exactly as other recognized attributes are.
2. **Publish/upgrade verification**: the IOTA bytecode verifier validates each function marked as a view function against the IIP-0011 signature constraints. A package that marks a non-conforming function as `#[view]` fails to publish — this is what lets readers trust the recorded list without re-checking signatures.
3. **Object creation**: for each module, the protocol writes the verified view function names into the module's `ModuleMetadata`.

Because the recorded list is protocol-attested, a reader can treat membership in the list as a trusted "this function satisfies the view constraints" claim.

### Conditional creation

> The conditional-creation rule from IIP-0010 (a `PackageMetadata` object is created only when meaningful metadata exists) is preserved and broadened to the new kinds: under this layout, a `PackageMetadataV1` is created when any module of the package contributes at least one kind of metadata (an authenticator **or** a view function). Modules with no recognized metadata are not recorded.

Reviewer: I'd skip this section, doesn't really add new information

> ### Activation

Reviewer: I'd remove this section. See https://github.com/iotaledger/IIPs/pull/48/changes#r3559778827

The `modules_metadata` inline field of `PackageMetadataV1` is retained for compatibility but is empty under the new layout and deprecated for reads.

## Test Cases

1. Authenticator metadata round-trips through the dynamic-field layout (publish a package with `#[authenticator]` functions; read back via both the new and the deprecated accessors).
2. View function metadata is recorded for `#[view]` functions and queryable via `is_view_function_v1` / `borrow_view_functions_metadata_v1`.
3. Conditional creation: a package with neither authenticators nor view functions produces no `PackageMetadata` object; a package with only view functions does.
4. A package upgrade that removes a view function produces an updated metadata object for the new version while the previous version's metadata stays valid.
5. Deprecated accessors behave identically against legacy inline-layout objects and dynamic-field-layout objects.
6. The bytecode verifier rejects a package that marks a non-conforming function as `#[view]` at publish time.

## Reference Implementation

> Implemented in the IOTA core monorepo. Key components:

Reviewer: Please, refer to the PR instead
